builtin/providers/aws/resource_aws_s3_bucket.go

```diff
@@ -201,17 +201,21 @@ func resourceAwsS3BucketCreate(d *schema.ResourceData, meta interface{}) error {
 	// Get the bucket and acl
 	bucket := d.Get("bucket").(string)
 	acl := d.Get("acl").(string)
+	region := d.Get("region").(string)
 
 	log.Printf("[DEBUG] S3 bucket create: %s, ACL: %s", bucket, acl)
 
 	req := &s3.CreateBucketInput{
 		Bucket: aws.String(bucket),
 		ACL:    aws.String(acl),
 	}
-
-	// Special case us-east-1 region and do not set the LocationConstraint.
-	// See "Request Elements: http://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketPUT.html
-	if awsRegion != "us-east-1" {
+	if *aws.String(region) != "" {
+		req.CreateBucketConfiguration = &s3.CreateBucketConfiguration{
+			LocationConstraint: aws.String(region),
+		}
+	} else if awsRegion != "us-east-1" {
+		// Special case us-east-1 region and do not set the LocationConstraint.
+		// See "Request Elements: http://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketPUT.html
```

Comment:
Hi Clint, thank you for replying here too! Your block looks good too and I'll update to this PR to use it.

The last peace to the puzzle before I remove [wip] to merge:
[`Regions outside of us-east-1 require the appropriate LocationConstraint to be specified in order to create the bucket in the desired region`](http://docs.aws.amazon.com/cli/latest/reference/s3api/create-bucket.html)

To clarify that, the default region is us-east-1. In the api debug, you can see the PUT URL doesn't need to include the region:
```http
PUT http://my-tf-test-bucket.s3.amazonaws.com/ HTTP/1.1
*empty body*
```
Compared to other regions:
```http
PUT http://my-tf-test-bucket.s3-us-west-1.amazonaws.com/ HTTP/1.1
*snip*
<CreateBucketConfiguration><LocationConstraint>us-west-1</LocationConstraint></CreateBucketConfiguration>
```

Adding a LocationConstraint of us-east-1 to `<bucket>.s3.amazonaws.com` (the implicit us-east-1), returns a 400. (that's the "Special case" seen in the code comments.) 
Specifying a region in the codified resource, other than what is set by `AWS_DEFAULT_REGION`, returns a `HTTP/1.1 307 Temporary Redirect`.

I believe the issue is the API URL is being set in the AWS golang SDK object by being initialized, `client.s3conn = s3.New(sess)` with the region from `AWS_DEFAULT_REGION`. This would break the ability to change the LocationConstraint to anything else.

Does that make sense? If so, what is the best way to restart the client session if we run into a resource region outside that specified by `AWS_DEFAULT_REGION`?